ntbs-service-unit-tests/Jobs/UpdateTableCountsJobTest.cs

```diff
@@ -0,0 +1,193 @@
+            _tableCountsRepository.Verify(r => r.ExecuteUpdateTableCountsStoredProcedure());
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        public async Task Run_ThrowsException_WhenNotEnoughRecords(int numberOfRecords)
+        {
+            // Arrange
+            _tableCountsRepository
+                .Setup(r => r.GetRecentTableCounts())
+                .Returns(Task.FromResult(Enumerable.Repeat(CreateBasicTableCounts(), numberOfRecords)));
+
+            // Act and assert
+            await Assert.ThrowsAsync<ApplicationException>(() => _updateTableCountsJob.Run(null));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        public async Task Run_LogsInfo_WhenNoFieldsReduced(int count)
+        {
+            // Arrange
+            _tableCountsRepository
+                .Setup(r => r.GetRecentTableCounts())
+                .Returns(Task.FromResult(Enumerable.Repeat(CreateBasicTableCounts(count), 2)));
+
+            Log.Logger = new LoggerConfiguration().WriteTo.TestCorrelator().CreateLogger();
+
+            using (TestCorrelator.CreateContext())
+            {
+                // Act
+                await _updateTableCountsJob.Run(null);
+
+                // Assert
+                var logEvents = TestCorrelator.GetLogEventsFromCurrentContext().ToList();
+
+                Assert.DoesNotContain(logEvents, logEvent => logEvent.Level > LogEventLevel.Information);
+                Assert.Contains(logEvents, logEvent => logEvent.Level == LogEventLevel.Information
+                                                       && logEvent.RenderMessage() == "Table counts look normal");
+            }
+        }
+
+        [Theory]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
+        [InlineData(1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)]
```

Comment:
This is an example of where reflection may be a good idea - we could write this a test that takes the name of the property that has decreased, sets the property value by reflection and then checks for failure. Doing it this way would also let us check more closely on the error message, which would be nice.

Reply:
Yep I've had a go at this, is it roughly what you had in mind?

Reply:
Yep, perfect :-)